js/src/data/api-fetch-middlewares.test.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Internal dependencies
+ */
+import { createErrorResponseCatcher } from './api-fetch-middlewares';
+import { API_NAMESPACE } from './constants';
+
+describe( 'createErrorResponseCatcher', () => {
+	// Ref: https://github.com/WordPress/gutenberg/blob/%40wordpress/api-fetch%405.1.1/packages/api-fetch/src/index.js#L68-L81
+	const checkStatus = ( response ) => {
+		if ( response.status >= 200 && response.status < 300 ) {
+			return response;
+		}
+
+		throw response;
+	};
+
+	const createFetchHandler = ( status, jsonParsedData ) => () => {
+		const response = {
+			status,
+			json() {
+				return Promise.resolve( jsonParsedData );
+			},
+		};
+		return Promise.resolve( response ).then( checkStatus );
+	};
+
+	const passReject = ( response ) => Promise.reject( response );
+
+	const optionsDefaultParse = { path: `${ API_NAMESPACE }/hi` };
+	const optionsDontParse = { path: `${ API_NAMESPACE }/hi`, parse: false };
+
+	describe( 'force overwrite `parse` option to `false` and return the same result', () => {
+		let middleware;
+
+		beforeEach( () => {
+			middleware = createErrorResponseCatcher( passReject );
+		} );
+
+		it( 'should only handle requests where the path belongs to this plugin', async () => {
+			const data = {};
+			const optionsOtherNamespace = { path: '' };
+			const callback = jest.fn().mockImplementation( ( options ) => {
+				expect( options ).toBe( optionsOtherNamespace );
+				return data;
+			} );
+
+			const result = await middleware( optionsOtherNamespace, callback );
+			expect( callback ).toHaveBeenCalledTimes( 1 );
+			expect( result ).toBe( data );
+		} );
+
+		it( 'should resolve parsed response body by default in a successful response', async () => {
+			const handler = createFetchHandler( 200, { data: 'hi' } );
+			const result = await middleware( optionsDefaultParse, handler );
+			expect( result ).toEqual( { data: 'hi' } );
+		} );
+
+		it( 'should reject parsed response body by default in a failed response', async () => {
+			const handler = createFetchHandler( 418, { message: 'oops' } );
+			const promise = middleware( optionsDefaultParse, handler );
+			await expect( promise ).rejects.toEqual( { message: 'oops' } );
+		} );
+
+		it( 'should resolve response instance when indicating `parse: false` in a successful response', async () => {
+			const handler = createFetchHandler( 200, { data: 'hi' } );
+			const result = await middleware( optionsDontParse, handler );
+			expect( result.status ).toEqual( 200 );
+			expect( result.json ).toBeInstanceOf( Function );
```

Comment:
This test case intends to assert that the response instance should resolve directly without any parse process in this middleware. I made a change https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/752411d4c5582485dd950f3526c1a609f63358a4 to check it directly.

And the real response instance is created and passed in from the outside of this middleware's scope. But I think we can rely on the implementation of `@wordpress/api-fetch` and believe it will pass the real response instance when the `parse` option is `false`.